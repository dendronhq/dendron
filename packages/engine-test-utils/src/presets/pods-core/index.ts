import JSON from "./json";
import MARKDOWN from "./markdown";

export const PODS_CORE = {
  JSON,
  MARKDOWN,
};

export const PODS_PRESETS = [{ name: "json", presets: JSON }];
