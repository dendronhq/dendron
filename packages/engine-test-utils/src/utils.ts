import {
  CONSTANTS,
  DendronConfig,
  VaultUtils,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings,
} from "@dendronhq/common-all";
import { readYAML } from "@dendronhq/common-server";
import { AssertUtils } from "@dendronhq/common-test-utils";
import { DConfig, Git } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";

export async function checkString(body: string, ...match: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

const getWorkspaceFolders = (wsRoot: string) => {
  const wsPath = path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME);
  const settings = fs.readJSONSync(wsPath) as WorkspaceSettings;
  return _.toArray(settings.folders);
};

export function checkVaults(opts: WorkspaceOpts, expect: any) {
  const { wsRoot, vaults } = opts;
  const configPath = DConfig.configPath(opts.wsRoot);
  const config = readYAML(configPath) as DendronConfig;
  expect(_.sortBy(config.vaults, ["fsPath", "workspace"])).toEqual(
    _.sortBy(vaults, ["fsPath", "workspace"])
  );
  const wsFolders = getWorkspaceFolders(wsRoot);
  expect(wsFolders).toEqual(
    vaults.map((ent) => {
      const out: WorkspaceFolderRaw = { path: VaultUtils.getRelPath(ent) };
      if (ent.name) {
        out.name = ent.name;
      }
      return out;
    })
  );
}

export class GitTestUtils {
  static async createRepoForWorkspace(wsRoot: string) {
    const git = new Git({ localUrl: wsRoot });
    await git.init();
    await git.add("dendron.yml");
    await git.commit({ msg: "init" });
  }

  /** Creates a "bare" git repository, to be used as the remote for a workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteCreate(remoteDir: string) {
    const git = new Git({ localUrl: remoteDir, bare: true });
    await git.init();
  }

  /** Adds a bare repository created with `createRemote` as the remote for the workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteAdd(wsRoot: string, remoteDir: string) {
    const git = new Git({ localUrl: wsRoot, remoteUrl: remoteDir });
    await git.remoteAdd();
    // Need to push to be able to set up remote tracking branch
    await git.push({ remote: "origin", branch: await git.getCurrentBranch() });
  }

  /** Set up a workspace with a remote, intended to be used when testing pull or push functionality.
   *
   * @param wsRoot Directory where the workspace will be stored.
   * @param remoteDir Directory where the remote will be stored. The workspace will pull and push to this remote.
   */
  static async createRepoForRemoteWorkspace(wsRoot: string, remoteDir: string) {
    await this.createRepoForWorkspace(wsRoot);
    await this.remoteCreate(remoteDir);
    await this.remoteAdd(wsRoot, remoteDir);
  }

  static async createRepoWithReadme(root: string) {
    const git = new Git({ localUrl: root });
    await git.init();
    const readmePath = path.join(root, "README.md");
    fs.ensureFileSync(readmePath);
    await git.add(".");
    await git.commit({ msg: "init" });
  }
}
