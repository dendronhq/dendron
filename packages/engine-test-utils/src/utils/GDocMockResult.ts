export const response = {
  "data": {
  "body": {
    "content": [
      {
        "endIndex": 1, 
        "sectionBreak": {
          "sectionStyle": {
            "contentDirection": "LEFT_TO_RIGHT", 
            "columnSeparatorStyle": "NONE", 
            "sectionType": "CONTINUOUS"
          }
        }
      }, 
      {
        "endIndex": 18, 
        "startIndex": 1, 
        "paragraph": {
          "elements": [
            {
              "endIndex": 18, 
              "startIndex": 1, 
              "textRun": {
                "content": "Testing GDoc Pod\n", 
                "textStyle": {}
              }
            }
          ], 
          "paragraphStyle": {
            "headingId": "h.11jz0sgxh4ph", 
            "direction": "LEFT_TO_RIGHT", 
            "namedStyleType": "HEADING_1"
          }
        }
      }, 
      {
        "endIndex": 41, 
        "startIndex": 18, 
        "paragraph": {
          "elements": [
            {
              "endIndex": 41, 
              "startIndex": 18, 
              "textRun": {
                "content": "This is the first line\n", 
                "textStyle": {}
              }
            }
          ], 
          "paragraphStyle": {
            "direction": "LEFT_TO_RIGHT", 
            "namedStyleType": "NORMAL_TEXT"
          }
        }
      }
    ]
  }, 
  "documentStyle": {
    "useCustomHeaderFooterMargins": true, 
    "pageSize": {
      "width": {
        "magnitude": 612, 
        "unit": "PT"
      }, 
      "height": {
        "magnitude": 792, 
        "unit": "PT"
      }
    }, 
    "marginHeader": {
      "magnitude": 36, 
      "unit": "PT"
    }, 
    "marginBottom": {
      "magnitude": 72, 
      "unit": "PT"
    }, 
    "marginFooter": {
      "magnitude": 36, 
      "unit": "PT"
    }, 
    "marginLeft": {
      "magnitude": 72, 
      "unit": "PT"
    }, 
    "background": {
      "color": {}
    }, 
    "marginRight": {
      "magnitude": 72, 
      "unit": "PT"
    }, 
    "pageNumberStart": 1, 
    "marginTop": {
      "magnitude": 72, 
      "unit": "PT"
    }
  }, 
  "title": "test", 
  "revisionId": "ALm37BXFqAKco_HUxV4VkthgGzmqZch7_jtkmDA7MPWbV5iDt-khIo2TIBbG_ANGhaGRXhMxR0JtdmMgfPzhOQ", 
  "documentId": "14o1AVg10CBbLlqNqBrCz4HkRNBmrwrXpuP5F___P_cw", 
  "suggestionsViewMode": "SUGGESTIONS_INLINE", 
  "namedStyles": {
    "styles": [
      {
        "textStyle": {
          "foregroundColor": {
            "color": {
              "rgbColor": {}
            }
          }, 
          "bold": false, 
          "baselineOffset": "NONE", 
          "strikethrough": false, 
          "weightedFontFamily": {
            "fontFamily": "Arial", 
            "weight": 400
          }, 
          "smallCaps": false, 
          "italic": false, 
          "fontSize": {
            "magnitude": 11, 
            "unit": "PT"
          }, 
          "backgroundColor": {}, 
          "underline": false
        }, 
        "namedStyleType": "NORMAL_TEXT", 
        "paragraphStyle": {
          "spacingMode": "COLLAPSE_LISTS", 
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "unit": "PT"
          }, 
          "keepWithNext": false, 
          "lineSpacing": 115, 
          "avoidWidowAndOrphan": true, 
          "borderBottom": {
            "color": {}, 
            "width": {
              "unit": "PT"
            }, 
            "dashStyle": "SOLID", 
            "padding": {
              "unit": "PT"
            }
          }, 
          "indentStart": {
            "unit": "PT"
          }, 
          "borderBetween": {
            "color": {}, 
            "width": {
              "unit": "PT"
            }, 
            "dashStyle": "SOLID", 
            "padding": {
              "unit": "PT"
            }
          }, 
          "borderRight": {
            "color": {}, 
            "width": {
              "unit": "PT"
            }, 
            "dashStyle": "SOLID", 
            "padding": {
              "unit": "PT"
            }
          }, 
          "borderLeft": {
            "color": {}, 
            "width": {
              "unit": "PT"
            }, 
            "dashStyle": "SOLID", 
            "padding": {
              "unit": "PT"
            }
          }, 
          "spaceAbove": {
            "unit": "PT"
          }, 
          "indentEnd": {
            "unit": "PT"
          }, 
          "shading": {
            "backgroundColor": {}
          }, 
          "borderTop": {
            "color": {}, 
            "width": {
              "unit": "PT"
            }, 
            "dashStyle": "SOLID", 
            "padding": {
              "unit": "PT"
            }
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "indentFirstLine": {
            "unit": "PT"
          }, 
          "alignment": "START", 
          "keepLinesTogether": false
        }
      }, 
      {
        "textStyle": {
          "fontSize": {
            "magnitude": 20, 
            "unit": "PT"
          }
        }, 
        "namedStyleType": "HEADING_1", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 6, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "magnitude": 20, 
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }, 
      {
        "textStyle": {
          "fontSize": {
            "magnitude": 16, 
            "unit": "PT"
          }, 
          "bold": false
        }, 
        "namedStyleType": "HEADING_2", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 6, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "magnitude": 18, 
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }, 
      {
        "textStyle": {
          "foregroundColor": {
            "color": {
              "rgbColor": {
                "blue": 0.2627451, 
                "green": 0.2627451, 
                "red": 0.2627451
              }
            }
          }, 
          "fontSize": {
            "magnitude": 14, 
            "unit": "PT"
          }, 
          "bold": false
        }, 
        "namedStyleType": "HEADING_3", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 4, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "magnitude": 16, 
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }, 
      {
        "textStyle": {
          "foregroundColor": {
            "color": {
              "rgbColor": {
                "blue": 0.4, 
                "green": 0.4, 
                "red": 0.4
              }
            }
          }, 
          "fontSize": {
            "magnitude": 12, 
            "unit": "PT"
          }
        }, 
        "namedStyleType": "HEADING_4", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 4, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "magnitude": 14, 
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }, 
      {
        "textStyle": {
          "foregroundColor": {
            "color": {
              "rgbColor": {
                "blue": 0.4, 
                "green": 0.4, 
                "red": 0.4
              }
            }
          }, 
          "fontSize": {
            "magnitude": 11, 
            "unit": "PT"
          }
        }, 
        "namedStyleType": "HEADING_5", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 4, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "magnitude": 12, 
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }, 
      {
        "textStyle": {
          "foregroundColor": {
            "color": {
              "rgbColor": {
                "blue": 0.4, 
                "green": 0.4, 
                "red": 0.4
              }
            }
          }, 
          "fontSize": {
            "magnitude": 11, 
            "unit": "PT"
          }, 
          "italic": true
        }, 
        "namedStyleType": "HEADING_6", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 4, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "magnitude": 12, 
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }, 
      {
        "textStyle": {
          "fontSize": {
            "magnitude": 26, 
            "unit": "PT"
          }
        }, 
        "namedStyleType": "TITLE", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 3, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }, 
      {
        "textStyle": {
          "foregroundColor": {
            "color": {
              "rgbColor": {
                "blue": 0.4, 
                "green": 0.4, 
                "red": 0.4
              }
            }
          }, 
          "fontSize": {
            "magnitude": 15, 
            "unit": "PT"
          }, 
          "weightedFontFamily": {
            "fontFamily": "Arial", 
            "weight": 400
          }, 
          "italic": false
        }, 
        "namedStyleType": "SUBTITLE", 
        "paragraphStyle": {
          "direction": "LEFT_TO_RIGHT", 
          "spaceBelow": {
            "magnitude": 16, 
            "unit": "PT"
          }, 
          "keepWithNext": true, 
          "spaceAbove": {
            "unit": "PT"
          }, 
          "namedStyleType": "NORMAL_TEXT", 
          "keepLinesTogether": true
        }
      }
    ]
  }
}
}

export const comments = {
  "data": {
  "items": [
    {
      "status": "open", 
      "kind": "drive#comment", 
      "author": {
        "picture": {
          "url": "//lh3.googleusercontent.com/a-/AOh14GjuAapwuLS18cloy_3ohiBaDFBetCREbPEMXf8g-4w=s50-c-k-no"
        }, 
        "kind": "drive#user", 
        "isAuthenticatedUser": true, 
        "displayName": "harshita joshi"
      }, 
      "deleted": false, 
      "replies": [
        {
          "kind": "drive#commentReply", 
          "author": {
            "picture": {
              "url": "//lh3.googleusercontent.com/a-/AOh14GjuAapwuLS18cloy_3ohiBaDFBetCREbPEMXf8g-4w=s50-c-k-no"
            }, 
            "kind": "drive#user", 
            "isAuthenticatedUser": true, 
            "displayName": "harshita joshi"
          }, 
          "deleted": false, 
          "htmlContent": "@abc This is a reply", 
          "content": "@abc This is a reply", 
          "replyId": "AAAANTbvk_0", 
          "modifiedDate": "2021-07-15T11:15:21.812Z", 
          "createdDate": "2021-07-15T11:15:21.812Z"
        }
      ], 
      "htmlContent": "first", 
      "content": "first", 
      "modifiedDate": "2021-07-15T11:15:21.812Z", 
      "context": {
        "type": "text/html", 
        "value": "line"
      }, 
      "createdDate": "2021-07-15T10:05:27.302Z", 
      "commentId": "AAAANcWdyCY", 
      "anchor": "kix.swkf5zd5ov94", 
      "fileTitle": "test", 
      "fileId": "sSjeCvERPVGU1NyfV6pdZxA"
    }
  ], 
  "kind": "drive#commentList", 
  "selfLink": "https://www.googleapis.com/drive/v2/files/14o1AVg10CBbLlqNqBrCz4HkRNBmrwrXpuP5F___P_cw/comments"
}
}