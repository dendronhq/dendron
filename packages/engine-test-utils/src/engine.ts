import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
