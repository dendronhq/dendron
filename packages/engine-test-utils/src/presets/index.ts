import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
