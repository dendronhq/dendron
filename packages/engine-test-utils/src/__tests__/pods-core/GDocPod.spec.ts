import { ENGINE_HOOKS } from "../../presets";
import { runEngineTestV5 } from "../../engine";
import { GDocImportPod } from "@dendronhq/pods-core";
import { VaultUtils } from "@dendronhq/common-all";
import {response, comments} from "../../utils/GDocMockResult";
import axios from 'axios';

jest.mock('axios');


describe("GDoc import pod", () => {
  let result: any;

  test("Import GDoc as Markdown", async () => {
    await runEngineTestV5(
      async ({ engine, vaults, wsRoot }) => {
        const pod = new GDocImportPod();
        const vaultName = VaultUtils.getName(vaults[0]);
        const mockedAxios = axios as jest.Mocked<typeof axios>;
        result=response;
        mockedAxios.get.mockResolvedValue(result);

        const { importedNotes } = await pod.execute({
          engine,
          vaults,
          wsRoot,
          config: {
            src: "foo",
            token: "xyzabcd",
            vaultName,
            documentId: "sdhdoj",
            hierarchyDestination: "gdoc.meet",
          },
        });
          expect(importedNotes[0].body).toMatch("\n\n## Testing GDoc Pod\n\nThis is the first line\n\n\n");
          
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
      }
    );
  });

  test("Import Comments in Markdown as json", async () => {
    await runEngineTestV5(
      async ({ engine, vaults, wsRoot }) => {
        const pod = new GDocImportPod();
        const vaultName = VaultUtils.getName(vaults[0]);
        const response = {
          fname: "gdoc.meet",
          custom: {
            documentId: 'sjkakauwu',
            revisionId: 'ALm37BXFqAKco_'
          },
          body: '\n\n## Testing GDoc Pod\n\nThis is the first line\n\n\n'
        };
        pod.getDataFromGDoc = jest.fn().mockReturnValue(response);
        const mockedAxios = axios as jest.Mocked<typeof axios>;
        result=comments;
        mockedAxios.get.mockResolvedValue(result);

        const { importedNotes } = await pod.execute({
          engine,
          vaults,
          wsRoot,
          config: {
            src: "foo",
            token: "xyzabcd",
            vaultName,
            documentId: "sdhdoj",
            hierarchyDestination: "gdoc.meet",
            importComments: {
              enable: true,
              format: "json"
            }
          },
        });
          expect(importedNotes[0].body).toMatchSnapshot()
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
      }
    );
  });

  test("Import Comments in Markdown as text", async () => {
    await runEngineTestV5(
      async ({ engine, vaults, wsRoot }) => {
        const pod = new GDocImportPod();
        const vaultName = VaultUtils.getName(vaults[0]);
        const response = {
          fname: "gdoc.meet",
          custom: {
            documentId: 'sjkakauwu',
            revisionId: 'ALm37BXFqAKco_'
          },
          body: '\n\n## Testing GDoc Pod\n\nThis is the first line\n\n\n'
        };
        pod.getDataFromGDoc = jest.fn().mockReturnValue(response);
        const mockedAxios = axios as jest.Mocked<typeof axios>;
        result=comments;
        mockedAxios.get.mockResolvedValue(result);

        const { importedNotes } = await pod.execute({
          engine,
          vaults,
          wsRoot,
          config: {
            src: "foo",
            token: "xyzabcd",
            vaultName,
            documentId: "sdhdoj",
            hierarchyDestination: "gdoc.meet",
            importComments: {
              enable: true,
              format: "text"
            }
          },
        });
          expect(importedNotes[0].body).toMatchSnapshot()
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
      }
    );
  });
});
