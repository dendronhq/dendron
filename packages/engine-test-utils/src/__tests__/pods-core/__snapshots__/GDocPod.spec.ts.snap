// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GDoc import pod Import Comments in Markdown as json 1`] = `
"

## Testing GDoc Pod

This is the first line


### Comments

 [{\\"author\\":\\"harshita joshi\\",\\"content\\":\\"first\\",\\"replies\\":[{\\"author\\":\\"harshita joshi\\",\\"content\\":\\"@abc This is a reply\\"}]}]"
`;

exports[`GDoc import pod Import Comments in Markdown as text 1`] = `
"

## Testing GDoc Pod

This is the first line


### Comments

 - harshita joshi:  first
	 - harshita joshi: @abc This is a reply
"
`;
